### PR TITLE
refactor: decouple DataTablesPane

### PR DESCRIPTION
**Pull request by zhaoyongjie**
### SUMMARY

In order to show the `results pane` in the Dashboard(separate [PR](https://github.com/apache/superset/pull/20144)), this PR decouples DataTablesPane and fix some bugs from the original implementation.

1. Samples Pane can't send a force query
2. Samples Pane can't re-query after editing the dataset
3. `Results Pane` and `Chart Pane` parallel sending queries.
4. correctly show the error message in `Results Pane` if it's an invalid query.
5. correctly show data in `Results Pane` and `samples Pane` when switching  dataset

### force-query Samples Pane

https://user-images.githubusercontent.com/2016594/169545812-a2c0c124-ab37-4419-833e-fd1b69be1f20.mov

### trigger re-query Samples Pane after editing the dataset

https://user-images.githubusercontent.com/2016594/169546257-fb99c565-931c-4fbf-bd6c-dccaae776bbd.mov

### parallel sending query
parallel query after

### ADDITIONAL INFORMATION

- [ ] Has associated issue:
- [ ] Required feature flags:
- [ ] Changes UI
- [ ] Includes DB Migration (follow approval process in [SIP-59](https://github.com/apache/superset/issues/13351))
  - [ ] Migration is atomic, supports rollback & is backwards-compatible
  - [ ] Confirm DB migration upgrade and downgrade tested
  - [ ] Runtime estimates and downtime expectations provided
- [ ] Introduces new feature or API
- [ ] Removes existing feature or API

### CODE REVIEW

File: superset-frontend/src/explore/components/DataTablesPane/useGetResultsOrSamples.ts [hunk 1]
```diff
@@ -0,0 +1,122 @@
...
+  isRequest,
+  resultType,
+  queryFormData,
+  datasource,
+  queryForce,
+  ownState,
+  actions,
+}: requestPayload): responseData {
+  // todo: this hook should split into useGetResults and useGetSamples after change endpoint
+  const [response, setResponse] = useState<responseData>({
```
Comment (kgabryje) [inline superset-frontend/src/explore/components/DataTablesPane/useGetResultsOrSamples.ts:37]: I think `useReducer` is a better choice for handling complex objects in a state. Alternatively, we could split that state into 5 `useState`s

Comment (zhaoyongjie) [reply superset-frontend/src/explore/components/DataTablesPane/useGetResultsOrSamples.ts]: done, I have added a new `samples` endpoint and refactored this logic.

File: superset-frontend/src/explore/components/DataTablesPane/components/ResultsPane.tsx [hunk 2]
```diff
@@ -0,0 +1,173 @@
...
+      getChartDataRequest({
+        formData: queryFormData,
+        force: queryForce,
+        resultFormat: 'json',
+        resultType: 'results',
+        ownState,
+      })
+        .then(({ json }) => {
+          const { colnames, coltypes } = json.result[0];
+          // Only displaying the first query is currently supported
```
Comment (villebro) [inline superset-frontend/src/explore/components/DataTablesPane/components/ResultsPane.tsx:70]: Idea for follow-up: Now it should be fairly simple to add support for displaying the other queries data, too. We could have a dropdown next to the search field where you could choose the query if there are more than one available.

File: superset-frontend/src/explore/components/DataTablesPane/DataTablesPane.tsx [hunk 3]
```diff
@@ -0,0 +1,214 @@
...
+  getItem,
+  setItem,
+  LocalStorageKeys,
+} from 'src/utils/localStorageHelpers';
+import { ResultsPane, SamplesPane, TableControlsWrapper } from './components';
+import { DataTablesPaneProps } from './types';
+
+enum ResultTypes {
+  results = 'results',
+  samples = 'samples',
```
Comment (villebro) [inline superset-frontend/src/explore/components/DataTablesPane/DataTablesPane.tsx:39]: nit, I believe this is the convention:
```suggestion
  Results = 'results',
  Samples = 'samples',
```

**Review by villebro — APPROVED** [review]
LGTM and tested to work well. I tried my best to compare some of the removed/added files, but it was difficult to get a good grasp of which functionality had changed, but everything looks sensible.

**Pull request merged**